### PR TITLE
Add gRPC support

### PR DESCRIPTION
**Pull request by ramindu90**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+    private void initiateSequenceAndInjectPayload(StreamObserver responseObserver,
+                                                  Event receivedEvent,
+                                                  org.apache.synapse.MessageContext msgCtx) throws AxisFault {
+        String msgPayload = receivedEvent.getPayload();
+        String sequenceName = receivedEvent.getHeadersMap().get(HEADER_MAP_SEQUENCE_PARAMETER_NAME);
+        log.debug(receivedEvent.getHeadersMap().toString());
+        SequenceMediator seq;
+        if (sequenceName != null) {
+            log.debug(sequenceName + " sequence, received via gRPC headers.");
```
Comment (Nirothipan) [inline components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java]: Shall we check whether debug log is enabled here

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            seq = (SequenceMediator) synapseEnvironment.getSynapseConfiguration().getSequence(injectingSeq);
+        }
+        msgCtx.setProperty(SynapseConstants.IS_INBOUND, true);
+        //validating the sequence
+        if (seq != null) {
+            if (!seq.isInitialized()) {
+                seq.init(synapseEnvironment);
+            }
+            seq.setErrorHandler(onErrorSeq);
+            log.debug("injecting received gRPC message to sequence : " + injectingSeq);
```
Comment (Nirothipan) [inline components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java]: Shall we check whether debug log is enabled here

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/Event.java [hunk 3]
```diff
@@ -0,0 +1,818 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (Nirothipan) [inline components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/Event.java]: Shall we please add the licence header for this generated file

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/InboundGrpcListener.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        String injectingSeq = params.getInjectingSeq();
+        String onErrorSeq = params.getOnErrorSeq();
+        SynapseEnvironment synapseEnvironment = params.getSynapseEnvironment();
+        String portParam = params.getProperties().getProperty(InboundGrpcConstants.INBOUND_ENDPOINT_PARAMETER_GRPC_PORT);
+        try {
+            log.warn("Exception occurred when getting " + InboundGrpcConstants.INBOUND_ENDPOINT_PARAMETER_GRPC_PORT +
+                    " property. Setting the port as " + DEFAULT_INBOUND_ENDPOINT_GRPC_PORT);
+            port = Integer.parseInt(portParam);
+        } catch (NumberFormatException e) {
+            port = DEFAULT_INBOUND_ENDPOINT_GRPC_PORT;
```
Comment (Nirothipan) [inline components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/InboundGrpcListener.java]: Shall we log some error  msg here and set default port

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/EventOrBuilder.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (Nirothipan) [inline components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/EventOrBuilder.java]: Shall we add the license header please

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/EventServiceGrpc.java [hunk 6]
```diff
@@ -0,0 +1,348 @@
+package org.wso2.micro.integrator.inbound.endpoint.protocol.grpc.util;
```
Comment (Nirothipan) [inline components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/EventServiceGrpc.java]: Shall we add the license header

File: components/mediation/mediators/grpc-reponse-mediator/org.wso2.micro.integrator.grpc.response.mediator/pom.xml [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                <configuration>
+                    <instructions>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Bundle-Description>${project.description}</Bundle-Description>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Export-Package>
+                            org.wso2.micro.integrator.grpc.response.mediator.*,
+                        </Export-Package>
+                        <Import-Package/>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (Nirothipan) [inline components/mediation/mediators/grpc-reponse-mediator/org.wso2.micro.integrator.grpc.response.mediator/pom.xml]: If possible, shall we define the needed imports

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java (hunk 1 above)
Comment (ramindu90) [reply components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java]: @Nirothipan fixed

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java (hunk 2 above)
Comment (ramindu90) [reply components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java]: @Nirothipan fixed

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/Event.java (hunk 3 above)
Comment (ramindu90) [reply components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/Event.java]: @Nirothipan fixed

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/InboundGrpcListener.java (hunk 4 above)
Comment (ramindu90) [reply components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/InboundGrpcListener.java]: @Nirothipan fixed

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/EventServiceGrpc.java (hunk 6 above)
Comment (ramindu90) [reply components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/EventServiceGrpc.java]: @Nirothipan fixed

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+import org.apache.synapse.core.SynapseEnvironment;
+import org.apache.synapse.mediators.base.SequenceMediator;
+import org.wso2.micro.integrator.inbound.endpoint.protocol.grpc.util.Event;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+
+import static org.wso2.micro.integrator.inbound.endpoint.protocol.grpc.InboundGrpcConstants.HEADER_MAP_SEQUENCE_PARAMETER_NAME;
+
+
```
Comment (sdkottegoda) [inline components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java]: Unnecessary new line

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+/**
+ * Inject gRPC message into the sequence
```
Comment (sdkottegoda) [inline components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java]: Doc comments need to end with a period

File: components/mediation/mediators/grpc-reponse-mediator/org.wso2.micro.integrator.grpc.response.mediator/pom.xml (hunk 7 above)
Comment (ramindu90) [reply components/mediation/mediators/grpc-reponse-mediator/org.wso2.micro.integrator.grpc.response.mediator/pom.xml]: fixed in https://github.com/wso2/micro-integrator/pull/772/commits/4b6e08be7026b345e624111d313d7544d5da34ef

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java (hunk 8 above)
Comment (ramindu90) [reply components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java]: @sdkottegoda  fixed in https://github.com/wso2/micro-integrator/pull/772/commits/760e17b1f81893f5301d6ee1e02680cb53a47255

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java (hunk 9 above)
Comment (ramindu90) [reply components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/GRPCInjectHandler.java]: @sdkottegoda  fixed in https://github.com/wso2/micro-integrator/pull/772/commits/760e17b1f81893f5301d6ee1e02680cb53a47255

File: components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/EventOrBuilder.java (hunk 5 above)
Comment (ramindu90) [reply components/mediation/inbound-endpoints/org.wso2.micro.integrator.inbound.endpoint/src/main/java/org/wso2/micro/integrator/inbound/endpoint/protocol/grpc/util/EventOrBuilder.java]: @Nirothipan fixed in https://github.com/wso2/micro-integrator/pull/772/commits/06f7426b0a48c85a38c3c29ba4cd3c52dccea7e5

**Pull request closed without merging**